### PR TITLE
Add circuit and wasm tests

### PR DESCRIPTION
**Pull request by azazak123**

### CODE REVIEW

File: plonk_prover/Cargo.toml [hunk 1]
```diff
@@ -31,10 +31,16 @@ getrandom = { version = "0.2.7", default-features = false, features = ["js"] }
...
+[dev-dependencies]
+rand = "0.8.5"
+
+[target.'cfg(target_arch = "wasm32")'.dev-dependencies]
+wasm-bindgen-test = "0.3.0"
 
 [features]
 default = ["std"]
-std = [
+std = []
```
Comment (armyhaylenko) [inline plonk_prover/Cargo.toml:42]: does it build in `no_std` if we have nothing in `std` feature?

File: plonk_prover/src/lib.rs [hunk 2]
```diff
@@ -11,17 +11,19 @@ use dusk_plonk::prelude::*;
 use dusk_poseidon::sponge;
 use rand_core::OsRng;
 
+///Constant which should be equal during generating proof and verifying its
 const TRANSCRIPT_INIT: &[u8; 7] = b"slushie";
 
+///Depth which is used in Slushie mixer contract
+pub const DEFAULT_DEPTH: usize = 20;
```
Comment (armyhaylenko) [inline plonk_prover/src/lib.rs:18]: i suggest creating a `shared` crate in the project root and putting such values over there; I suspect there is going to be more stuff to add there

Comment (armyhaylenko) [reply plonk_prover/src/lib.rs:18]: And this crate will be used by both Slushie and `plonk_prover`

File: plonk_prover/src/lib.rs [hunk 3]
```diff
@@ -267,4 +273,351 @@ mod tests {
...
+
+    ///Test for checking circuit works with random arguments
+    #[test]
+    fn random_args() {
+        const DEPTH: usize = 24;
+
+        // All arguments are random
+        let k = rand::random::<u32>();
+        let r = rand::random::<u32>();
+        let l = rand::random::<u16>() as usize;
```
Comment (armyhaylenko) [inline plonk_prover/src/lib.rs:383]: doesn't `random::<usize>()` work?

File: plonk_prover/src/lib.rs [hunk 4]
```diff
@@ -267,4 +273,351 @@ mod tests {
...
+            BlsScalar::from(f).into(),
+        ];
+
+        SlushieCircuit::<DEPTH>::verify(&pp, &dv, &proof.unwrap(), &public_inputs, TRANSCRIPT_INIT)
+            .unwrap();
+    }
+
+    ///Test for checking circuit works with wrong fee
+    #[test]
+    #[should_panic]
```
Comment (armyhaylenko) [inline plonk_prover/src/lib.rs:406]: consider adding the exact error here

File: plonk_prover/src/lib.rs [hunk 5]
```diff
@@ -267,4 +273,351 @@ mod tests {
...
+            BlsScalar(bytes_to_u64(R)).into(),
+            sponge::hash(&[(k as u64).into()]).into(),
+            BlsScalar::from_raw(bytes_to_u64(PAYOUT)).into(),
+            BlsScalar::from_raw(bytes_to_u64(RELAYER)).into(),
+            // Fee in public inputs is incorrect
+            BlsScalar::from(f + 1).into(),
+        ];
+
+        SlushieCircuit::<DEPTH>::verify(&pp, &dv, &proof.unwrap(), &public_inputs, TRANSCRIPT_INIT)
+            .unwrap();
```
Comment (armyhaylenko) [inline plonk_prover/src/lib.rs:430]: consider replacing this `unwrap()` with `expect()`

File: plonk_prover/src/lib.rs [hunk 6]
```diff
@@ -267,4 +273,351 @@ mod tests {
...
+    }
+
+    ///Test for checking circuit works with wrong nullifier hash
+    #[test]
+    #[should_panic]
```
Comment (armyhaylenko) [inline plonk_prover/src/lib.rs:435]: consider adding the exact error here

**Review by armyhaylenko — COMMENTED** [review]
Good work! Left a couple of comments.
Adding `#[should_panic = "your error"]` is desirable. Adding `expect()` is optional, but please do it if it gives us more control over the error messages and makes them better

File: plonk_prover/src/lib.rs (hunk 3 above)
Comment (azazak123) [reply plonk_prover/src/lib.rs:405]: For tests, I use different depth. For example, if depth = 20, then `0 <= leaf_index <= 2^depth - 1`, but if `leaf_index = random::<usize>()` than `0 <= leaf_index <= 2^(pointer size in bits) - 1`. In this way, we should use u(number), where `number <= depth`

File: plonk_prover/src/circuit.rs [hunk 7]
```diff
@@ -1,5 +1,7 @@
 use crate::*;
 
+use alloc::vec;
+use alloc::vec::Vec;
```
Comment (armyhaylenko) [inline plonk_prover/src/circuit.rs:4]:
```suggestion
use alloc::vec::{self, Vec};
```

**Review by armyhaylenko — APPROVED** [review]
LGTM

**Review by armyhaylenko — APPROVED** [review]
LGTM!

**Review by goodelita1 — APPROVED** [review]

**Review by armyhaylenko — APPROVED** [review]

**Review by goodelita1 — APPROVED** [review]

**Pull request merged**